File: examples/python/input/history.py
#!/usr/bin/env python3
"""Input history with Up/Down recall and persistence.

    make run-example EX=python/input/history

As a context manager, History saves on exit, so a second run recalls lines
from the first.
"""

from __future__ import annotations

import sparcli as sc

APP_NAME = 'sparcli-history-example'


def main() -> None:
    if not sc.input_available():
        sc.alert.warning('Run this example in a real terminal (not piped).')
        return

    # app=... persists to ~/.local/state/<app>/history; loaded on creation,
    # saved on context-manager exit.
    with sc.History(app=APP_NAME, max_entries=100) as history:
        if len(history) > 0:
            sc.println(f'Loaded {len(history)} line(s) from the previous run; '
                       'press Up to recall them.')

        # Submitted lines are auto-added (empty + duplicates skipped).
        for _ in range(3):
            line = sc.text_input('>', sc.TextInputOpts(
                history=history, placeholder='type, or press Up'))
            if line is None:
                break

        sc.println('\nHistory contents:')
        for i, entry in enumerate(history.entries()):
            sc.println(f'  {i:2}: {entry}')


if __name__ == '__main__':
    main()

File: examples/rust/input/history.rs
//! Input history with Up/Down recall and persistence.
//!
//!   cargo run -p sparcli --example input_history
//!
//! Dropping the `History` saves the configured file, so a second run recalls
//! lines from the first.

use sparcli::*;

const APP_NAME: &str = "sparcli-history-example";

fn main() -> sparcli::Result<()> {
    if !input_available() {
        alert::warning("Run this example in a real terminal (not piped).");
        return Ok(());
    }

    // `.app(...)` persists to ~/.local/state/<app>/history; loaded on new().
    let history = History::new(HistoryOpts::new().app(APP_NAME).max_entries(100));

    if history.count() > 0 {
        println(
            &format!(
                "Loaded {} line(s) from the previous run; press Up to recall.",
                history.count()
            ),
            Style::default(),
        );
    }

    // Submitted lines are auto-added (empty + consecutive duplicates skipped).
    for _ in 0..3 {
        match text_input(">", TextInputOpts::new().history(&history).placeholder("type, or press Up")) {
            Ok(Some(_)) => {}
            _ => break,
        }
    }

    println("\nHistory contents:", Style::default());
    for i in 0..history.count() {
        println(
            &format!("  {i:2}: {}", history.get(i).unwrap_or_default()),
            Style::default(),
        );
    }

    // Dropping `history` saves it to the state file.
    Ok(())
}
